### ISSUE TITLE
[Question] Is that possible to implement operator with Julia?

### ISSUE THREAD

**Issue by jingpengwu** [malmaud/TensorFlow.jl#181]
Thanks for this great package. 
Are all the operators need to be implemented in C++?

**Comment by malmaud** [comment]
Thanks!
For now, they do. Google plans to eventually enable creating new operations on the fly from C (see https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/c_api.h#L922), at which point it should be possible to define new operations in Julia.

**Comment by jingpengwu** [comment]
cool! thanks for the quick response. we can keep this issue here to track this capability?

**Comment by malmaud** [comment]
Sure.

**Comment by oxinabox** [comment]
I don't overly see the point in being able to create operators directly in Julia.

TensorFlow is more than turing complete;
any new operators one desires can easily be implemented inside Julia, out of the parts we already have.
(Or if required out of parts that we can add from the C API)

See for example [here](https://github.com/oxinabox/ColoringNames.jl/blob/05cfb55588f82a640f8bc19c1f5ec98305111e9b/src/col2term_bigram.jl#L5-L23)

In a practical sense we can define all the operations we want.
The main advantage that I can see to defining something as a proper operation,
rather than just building it from parts,
is that it will be accessible too all language bindings of tensorflow.
But that only applies if it is in the C API.
And so more or less need to be written in C++.

**Comment by jingpengwu** [comment]
@oxinabox your example is pretty cool, but there is only one function in the session. Is that possible to compose multiple Julia functions with/without TF functions and use TF to parallelly schedule functions?

**Comment by oxinabox** [comment]
@jingpengwu I'm not sure i understand the question.

**Comment by jingpengwu** [comment]
@oxinabox Here is an [example](https://github.com/oxinabox/ColoringNames.jl/blob/e053f6f6f32fa622017d03873cd676be9353776c/test/LSTM.jl#L15)
There is only one Julia function in the TensorFlow session, is that possible to add more functions? After doing so, can we expect that TensorFlow will build a computational graph of Julia functions, and explore parallelism inside the compuational graph.

**Comment by oxinabox** [comment]
It isn't exactly building a computational graph of Julia functions.
The julia functions (like for all operations available in TensorFlow.jl) return Tensors.
So they are building the Computational Graph, which is given to the C tensorflow library to execute.

You can certainly use many julia functions that return Tensors as part of one graph.
Eg [here](https://github.com/oxinabox/ColoringNames.jl/blob/05cfb55588f82a640f8bc19c1f5ec98305111e9b/src/col2term_bigram.jl#L79-L85)
The graph doesn't care how it is built.

**Comment by jingpengwu** [comment]
Your functions are still based on TF functions, such as `gather_nd`, to manipulate the computation graph. Can we use normal Julia function, such as `apply_mask(V, mask) = V.*mask`, as an node of the computation graph?

**Comment by oxinabox** [comment]
> Can we use normal Julia function, such as `apply_mask(V, mask) = V.*mask`, as an node of the computation graph?

Generally, no.
And as I said before I'm not sure what it would be useful for. (I'm not sure it wouldn't be useful)

Consider that  `V.*mask` is in fact still a TF function, as you are calling it.
It is define around [here](https://github.com/malmaud/TensorFlow.jl/blob/e636f565d7453310e5d4320c49d27b2779a0e2f2/src/ops/math.jl#L91-L123)
It is a wrapper around the Tensorflow C definition for `multiply`.

I don't think we are too far from the point where there is enough of this kind of thing, that it it becomes really hard to differentiate functions that are for julia on AbstractArrays, and functions that are for TensorFlow on Tensors

**Comment by oxinabox** [comment]
Oh does this mean:
[this](https://www.tensorflow.org/extend/adding_an_op)

I think, maybe, just maybe, this could be done somehow via CXX.jl?

**Comment by jingpengwu** [comment]
yep, it is adding an operator, but based on pure Julia rather than C++.

**Comment by stevengj** [comment]
Exactly.  Once you can define an operator that calls back to an arbitrary pure-Julia function, then you potentially get a whole bunch of things (like fusing broadcasts) for free.

Note that in principle, you may only need to define *one* C++ op (or use CXX.jl) that stores a handle to a Julia callback function as internal state, and maybe another function to (optionally) compute the gradient.     Then each time you have a new Julia function, you just instantiate a new instance of the Op and pass the Julia function in its constructor.

**Comment by stevengj** [comment]
A complication is that `REGISTER_OP` in TensorFlow requires you to specify the input and output types, which doesn't map well onto Julia functions (that may allow multiple types).  However, you can at least register callback-based ops for a few common cases (integers or floating-point values in and out).

**Comment by malmaud** [comment]
There are a few practical problems with defining an atomic TensorFlow operator for arbitrary Julia functions:

* We would need to distribute custom TensorFlow binaries if we want to use custom ops, until CXX.jl is supported robustly across everyone's Julia distributions.
* AFAIK, there wouldn't be a way to plug in Julia callbacks to TensorFlow's GPU or TPU backend, which is the majority of the use-case.
* Most users of TensorFlow.jl will want automatic gradients for implementing gradient descent. TensorFlow gives you automatic gradients for free when you represent a computation as a composite of TensorFlow operations; if you only have a single atomic TensorFlow operator with a Julia callback, then all the responsibility of automatic differentiation falls to TensorFlow.jl.

**Comment by stevengj** [comment]
Regarding gradients, with custom operations my understanding is that TensorFlow allows you to supply a gradient function too.   e.g. you could automate this with ForwardDiff applied to the broadcast operand.   Then the rest of TensorFlow's gradient machinery would work when this operation is composed with other TensorFlow operations.
